src/prefect_server/api/flows.py

```diff
@@ -194,11 +205,39 @@ async def create_flow(
                 "heartbeat_enabled": True,
                 "lazarus_enabled": True,
                 "version_locking_enabled": False,
+                "idempotency_key": idempotency_key,
             },
         ).insert()
     else:
         flow_group_id = flow_group.id
 
+        # check idempotency key and  early if we find a matching key and flow,
+        # otherwise update the key for the group
+        last_idempotency_key = flow_group.settings.get("idempotency_key", None)
+        if (
+            last_idempotency_key
+            and idempotency_key
+            and last_idempotency_key == idempotency_key
+        ):
+            # get the most recent unarchived version, there should only be one
+            # unarchived flow at a time but it is safer not to presume
+            flow = await models.Flow.where(
+                {
+                    "version_group_id": {"_eq": version_group_id},
+                    "archived": {"_eq": False},
+                }
+            ).first(order_by={"version": EnumValue("desc")})
+            if flow:
+                return flow.id
+            # otherwise, despite the key matching we don't have a valid flow to return
+            # and will continue as though the key did not match
+
+        settings = flow_group.settings
+        settings["idempotency_key"] = idempotency_key
+        await models.FlowGroup.where({"id": {"_eq": flow_group.id}}).update(
+            set={"settings": settings}
+        )
```

Comment:
This update should only be triggered if `idempotency_key` is not `None`

Reply:
I'm not sure. That would allow keys to linger from past registrations so you could do something unexpected like
```
create_flow(key="foo") -> creates
create_flow(key=None) -> creates
create_flow(key="foo") -> does not create
```

Reply:
Oh yep my bad, I understood after I wrote this that `None` is essentially an idempotency key in this sense where it only matter what the most recent create call was passing 👍